vac/raw/decentralized-messaging-ethereum.md

```diff
@@ -0,0 +1,737 @@
+assigning it to `sk`. 
+If the secret key corresponds to one generated by ourselves, 
+that secret key along with all keys with lower index are deleted. 
+This deletion is indicated by `sigma.mySks[≤ keyIndex] = empty_string`. 
+Subsequently, the new public and secret keys contained in the message are stored.
+
+```
+(ciphertext, keySender, keyIndex) = c
+if keySender = "other" then 
+sk = sigma.mySks[keyIndex] 
+sigma.mySks[≤ keyIndex] = empty_string
+else sk = sigma.receivedSk
+(m, sigma.receivedSk, sigma.otherPkIndex, sigma.otherPk) = PKE-Dec(sk, ciphertext)
+sigma.otherPkSender = "other"
+return (sigma, m)
+```
+### PKE Syntax
+
+The required PKE that MUST be used is ElGamal with a 2048-bit modulus `p`. 
+
+#### Parameters
+
+The following parameters must be used:
+
+```
+p = 308920927247127345254346920820166145569
+g = 2
+```
+
+#### PKE-KGen
+
+Each user `u` MUST do the following:
+
+```
+PKE-KGen():
+a = randint(1, p-2)
+pk = (p, g, g^a)
+sk = a
+return (pk, sk)
+```
+
+#### PKE-Enc
+
+A user `v` encrypting a message `m` for `u` MUST follow these steps:
+
+```
+PKE-Enc(pk):
+k = randint(1, p-2)
+eta = g^k % p
+delta = m * (g^a)^k % p
+return ((eta, delta))
+```
+
+#### PKE-Dec
+
+The user `u` recovers a message `m` from a ciphertext `c` by performing the following operations:
+
+```
+PKE-Dec(sk):
+mu = eta^(p-1-sk) % p
+return ((mu` * delta) % p)
+```
+
+### DCGKA Syntax
+#### Auxiliary functions
+
+There exist 6 functions that are auxiliary for the rest of components of the protocol, namely:
+
+##### init
+
+This function takes an `ID` as input and returns its associated initial state, denoted by `gamma`:
+
+```
+gamma.myId = ID
+gamma.mySeq = 0
+gamma.history = empty
+gamma.nextSeed = empty_string
+gamma.2sm[·] = empty_string
+gamma.memberSecret[·, ·, ·] = empty_string
+gamma.ratchet[·] = empty_string
+return (gamma)
+```
+
+#### encrypt-to
+
+Upon reception of the recipient’s `ID` and a plaintext, 
+it encrypts a direct message for another group member. 
+Should it be the first message for a particular `ID`, 
+then the `2SM` protocol state is initialized and stored in `gamma.2sm[recipient.ID]`. 
+One then uses `2SM_Send` to encrypt the message and store the updated protocol in `gamma`.
+
+```
+if gamma.2sm[recipient_ID] = empty_string then
+gamma.2sm[recipient_ID] = 2SM_Init(gamma.myID, recipient_ID)
+(gamma.2sm[recipient_ID], ciphertext) = 2SM_Send(gamma.2sm[recipient_ID], plaintext)
+return (\gamma, ciphertext)
+```
+
+#### decrypt-from
+
+After receiving the sender’s `ID` and a ciphertext, 
+it behaves as the reverse function of `encrypt-to` and has a similar initialization:
+
+```
+if gamma.2sm[sender_ID] = empty_string then
+gamma.2sm[sender_ID] = 2SM_Init(gamma.myID, sender_ID)
+(gamma.2sm[sender_ID], plaintext) = 2SM_Receive(gamma.2sm[sender_ID], ciphertext)
+return (gamma, plaintext)
+```
+
+#### update-ratchet
+
+This function generates the next update secret `I` for the group member `ID`. 
```

Comment:
In this section I suggest to add an index for `update secret` value (e.g `I_j` or `I_next` or sth else). I think It would give better understanding for those one who go through function description that it is the same value as in start of rfc where it has indexes